README.md
# http-proxy-to-socks

Fork of https://github.com/oyyd/http-proxy-to-socks.

hpts(http-proxy-to-socks) is a nodejs tool to convert SOCKS proxy into http proxy.

Many clients support setting up http proxy to speed up network requests and for sometimes only SOCKS proxy is available to you. SOCKS proxy supports TCP so that it's possible to convert those requests from http proxy into SOCKS protocol. In this way, you can still keep the goodness provided by your SOCKS proxy(e.g. encryption).

This fork additionally provides the functionality to proxy only selected hosts and also define different socks proxies chosen by the target hostname of your request.

## Setup

This fork was not released on npm, so you have to install it manually.

1. if you have hpts from official npm installed run `npm uninstall -g http-proxy-to-socks`, to uninstall it. Run `hpts` to test whether it was removed correctly.
2. clone or download this repository
3. open a terminal in the downloaded directory
4. run `npm link` (will install dependencies and link the hpts command)
5. run the module using `hpts [parameters]` (from anywhere on your machine)

Make sure your nodejs version is greater/equal to `14`.

## Usage

```
hpts -s 127.0.0.1:1080 -p 8080
```

This will start a process listening on `8080` as a http proxy. It will convert http requests into socks requests and send them to port `1080`. Please make sure your socks service is available at the corresponding port.

Other options:

```
Options:

  -h, --help                  output usage information
  -V, --version               output the version number
  -s, --socks [socks]         specify your socks proxy host (only one possible through cli), default: 127.0.0.1:1080
  -l, --host [host]           specify the listening host of http proxy server, default: 127.0.0.1
  -p, --port [port]           specify the listening port of http proxy server, default: 8080
  -c, --config [config]       path to a json config file or the json config as string
  --level [level]             log level, vals: info, error, default: error
```

### Configuration File

The cli parameters do not support white-/blacklists. This is only possible using the `json` config file with `-c`:

```json
{
  "host": "127.0.0.1",
  "port": 8080,
  "level": "info",
  "proxies": [
    {
      "socks": "127.0.0.1:1080",
      "whitelist": [ "google\\.com", "\\.org$" ]
    },
    {
      "socks": "127.0.0.1:1081",
      "blacklist": [ "^github\\.com$" ]
    }
  ]
}
```

### White- and Blacklists

White- and blacklists can be used to determine hosts that either should or should not be proxied. These are regular expressions being matched against the **hostname** of the target. The protocol/host/path are not part of this matching.
Make sure to use correct RegExp syntax für JavaScript. You can only have either a white- or a blacklist. Since a whitelist is usually more limiting, if both are specified only the whitelist will be used.

Hosts defined in a whitelist will be proxied, all other hosts will be accessed without going through the socks proxy defined. Blacklisted vice versa.

## CONTRIBUTE

Please add more tests for corresponding features when you send a PR:

```
npm run test
```

## License

MIT


CHANGELOG.md
# CHANGELOG

## 2.0.0
- completely renewed dependencies and removed unnecessary overhead
- implemented new configuration properties to allow different socks proxies at the same time
- added white-/blacklisting based on target hostnames

## 1.1.1
- src: Use parsed url hostname instead of host which may contain also port

## 1.1.0
- src: Support setting local address of the http-service.
- src: Format timestamp of logs.

## 1.0.4
- src: Support setting the `--level` option in `createServer()`.

## 1.0.3
- src: Bind `error` event on sockets to prevent servers from exiting in some node runtimes.

## 1.0.2

- Exports "main" fields in `package.json`.


src/__tests__/cli.spec.js
const { getOptionsArgs } = require('../cli');

describe('cli', () => {
  describe('getOptionsArgs', () => {
    it('should return an object with specified fields when provided', () => {
      const options = {
        socks: 'socks',
        port: 'port',
        level: 'level',
        config: 'config',
        FIELD_NOT_EXIST: 'FIELD_NOT_EXIST',
      };

      const res = getOptionsArgs(options);

      Object.keys(options).forEach((name) => {
        if(name === 'FIELD_NOT_EXIST') {
          expect(res[name]).toBeFalsy();
        } else if(name === 'socks') {
          expect(res.proxies[0]).toBeTruthy();
        } else {
          expect(res[name]).toBeTruthy();
        }

      });
    });
  });
});


src/__tests__/proxy_server.spec.js
const mockAgent = {};

const mockOn = jest.fn();

jest.mock('http', () => ({
  request: jest.fn(() => ({
    on: mockOn,
  })),
  Server: jest.fn(function Server() {
    this.addListener = jest.fn();
  }),
}));

jest.mock('socks', () => ({
  createConnection: jest.fn(),
  Agent: jest.fn(() => mockAgent),
}));

function last(array) {
  return array[array.length - 1];
}

function getLastMockOn(event) {
  return last(mockOn.mock.calls.filter(args => args[0] === event));
}

const http = require('http');
const Socks = require('socks');
const {
  createServer,
  getProxyObject,
  parseProxyLine,
  requestListener,
  connectListener,
  getProxyInfo,
} = require('../proxy_server');

describe('proxy_server', () => {
  const requestURL = 'https://google.com';
  let proxyList;
  let request;
  let socksRequest;
  let response;
  let socketRequest;
  let socket;

  beforeEach(() => {
    proxyList = [{
      socksProxy: {
        ipaddress: '127.0.0.1',
        port: 8080,
        type: 5,
        authentication: { username: '', password: '' },
      },
      whitelist: [
        new RegExp('google\\.com'),
        new RegExp('\\.org$')
      ]
    },
    {
      socksProxy: {
        ipaddress: '127.0.0.1',
        port: 8081,
        type: 5,
        authentication: { username: '', password: '' },
      },
      blacklist: [
        new RegExp('^github\\.com$')
      ]
    }
    ];

    request = {
      on: jest.fn(),
      pipe: jest.fn(),
      url: requestURL,
    };

    socksRequest = {
      on: jest.fn(),
      pipe: jest.fn(),
      url: requestURL.slice('https://'.length),
    };

    response = {
      on: jest.fn(),
      writeHead: jest.fn(),
      end: jest.fn(),
    };

    socketRequest = {
      on: jest.fn(),
      write: jest.fn(),
      pipe: jest.fn(),
    };

    socket = {
      on: jest.fn(),
      pipe: jest.fn(),
      write: jest.fn(),
      resume: jest.fn(),
    };
  });

  describe('getProxyObject', () => {
    it('should return an object with "ipaddress", "port", "type", "authentication" properties', () => {
      const host = '127.0.0.1';
      const port = '8080';
      const res = getProxyObject(host, port);

      expect(typeof res).toBe('object');
      expect(res.ipaddress).toBe(host);
      expect(res.port).toBe(parseInt(port, 10));
      expect(res.type).toBe(5);
      expect(typeof res.authentication).toBe('object');
      expect(Object.hasOwnProperty.apply(res.authentication, ['username'])).toBeTruthy();
      expect(Object.hasOwnProperty.apply(res.authentication, ['password'])).toBeTruthy();
    });
  });

  describe('parseProxyLine', () => {
    it('should return a object with "host" and "port" extracted from proxy string', () => {
      const proxyLine = '127.0.0.1:1080';
      const res = parseProxyLine(proxyLine);

      expect(typeof res).toBe('object');
      expect(res.ipaddress).toBe('127.0.0.1');
      expect(res.port).toBe(1080);
    });

    it('should also contain "username" and "password" properties when it contains these info', () => {
      const proxyLine = '127.0.0.1:1080:oyyd:password';
      const res = parseProxyLine(proxyLine);

      expect(typeof res).toBe('object');
      expect(res.ipaddress).toBe('127.0.0.1');
      expect(res.port).toBe(1080);
      expect(typeof res.authentication).toBe('object');
      expect(res.authentication.username).toBe('oyyd');
      expect(res.authentication.password).toBe('password');
    });

    it('should throw error when the proxy string seems not good', () => {
      let proxyLine = '127.0.0.1';
      let error = null;

      try {
        error = parseProxyLine(proxyLine);
      } catch(err) {
        error = err;
      }

      expect(error instanceof Error).toBeTruthy();

      proxyLine = '127.0.0.1:8080:oyyd';
      error = null;

      try {
        error = parseProxyLine(proxyLine);
      } catch(err) {
        error = err;
      }

      expect(error instanceof Error).toBeTruthy();
    });
  });

  describe('requestListener', () => {
    it('should create an socks agent and take it as request agent', () => {
      requestListener(proxyList, request, response);

      const lastCall = last(Socks.Agent.mock.calls);
      const httpLastCall = last(http.request.mock.calls);

      expect(requestURL.indexOf(lastCall[0].target.host) > -1).toBeTruthy();
      expect(httpLastCall[0].agent === mockAgent).toBeTruthy();
    });

    it('should return 500 when error thrown', () => {
      requestListener(proxyList, request, response);

      const onErrorArgs = getLastMockOn('error');

      expect(onErrorArgs).toBeTruthy();

      const error = new Error('500');

      onErrorArgs[1](error);

      expect(response.writeHead.mock.calls[0][0]).toBe(500);
      expect(response.end.mock.calls[0][0].indexOf('error') > -1).toBeTruthy();
    });

    it('should pipe response when "response"', () => {
      const proxyResponse = {
        statusCode: 200,
        headers: {},
        pipe: jest.fn(),
      };

      requestListener(proxyList, request, response);

      const onResponseArgs = getLastMockOn('response');

      expect(onResponseArgs).toBeTruthy();

      onResponseArgs[1](proxyResponse);

      expect(proxyResponse.pipe.mock.calls[0][0]).toBe(response);
    });
  });

  describe('connectListener', () => {
    it('should create socks connections', () => {
      const head = '';
      connectListener(proxyList, socksRequest, socketRequest, head);

      const lastCreateConnectionCall = last(Socks.createConnection.mock.calls);

      expect(lastCreateConnectionCall[0].target.host).toBe('google.com');
    });

    it('should write 500 when error thrown', () => {
      const head = '';
      connectListener(proxyList, socksRequest, socketRequest, head);

      const lastCreateConnectionCall = last(Socks.createConnection.mock.calls);

      const error = new Error('500');

      lastCreateConnectionCall[1](error, socket);

      expect(socketRequest.write.mock.calls[0][0].indexOf('500') > -1).toBeTruthy();
      expect(socket.pipe.mock.calls.length === 0).toBeTruthy();
    });

    it('should pipe sockets when socket connected', () => {
      const head = '';

      connectListener(proxyList, socksRequest, socketRequest, head);

      const lastCreateConnectionCall = last(Socks.createConnection.mock.calls);

      lastCreateConnectionCall[1](null, socket);

      expect(socketRequest.pipe.mock.calls[0][0]).toBe(socket);
      expect(socket.pipe.mock.calls[0][0]).toBe(socketRequest);
      expect(socketRequest.write.mock.calls[0][0].indexOf('200') > -1).toBeTruthy();
      expect(socket.write.mock.calls[0][0]).toBe(head);
    });
  });

  describe('createServer', () => {
    it('should push this.proxyList', () => {
      const options = {
        proxies: [{
          socks: '127.0.0.1:1080'
        }]
      };

      createServer(options);

      const { proxyList } = http.Server.mock.instances[0];

      expect(proxyList[0].socksProxy.ipaddress).toBe('127.0.0.1');
      expect(proxyList[0].socksProxy.port).toBe(1080);
    });

    it('should listen both "request" and "connect" events', () => {
      const options = {
        proxy: '127.0.0.1:1080',
      };

      createServer(options);

      const { addListener } = http.Server.mock.instances[0];

      const onRequestArgs = addListener.mock.calls.filter(args => args[0] === 'request');
      const onConnectArgs = addListener.mock.calls.filter(args => args[0] === 'connect');

      expect(onRequestArgs.length > 0).toBeTruthy();
      expect(onConnectArgs.length > 0).toBeTruthy();
    });
  });

  describe('getProxyInfo', () => {
    it('should select socks proxy by whitelist', () => {

      let result = getProxyInfo(proxyList, new URL('http://google.com'));

      expect(result.ipaddress).toBe('127.0.0.1');
      expect(result.port).toBe(8080);
    });

    it('should select socks proxy by avoiding blacklist', () => {

      let result = getProxyInfo(proxyList, new URL('http://google.de'));

      expect(result.ipaddress).toBe('127.0.0.1');
      expect(result.port).toBe(8081);
    });

    it('should return empty if no proxy matches', () => {

      let result = getProxyInfo(proxyList, new URL('http://github.com'));

      expect(result).toBe(undefined);
    });
  });
});


src/logger.js
const winston = require('winston');
const chalk = require('chalk');

const { format, createLogger: createWinstonLogger, transports } = winston;
const { Console } = transports;
const { combine, printf } = format;

const DEFAULT_COMMON_OPTIONS = {
  colorize: true,
  timestamp: true,
};

function fillDigit(num) {
  const str = String(num);
  if(str.length === 1) {
    return `0${str}`;
  }

  return str;
}

function getFormatedDate() {
  const date = new Date();

  return `${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()}`
    + ` ${fillDigit(date.getHours())}:${fillDigit(date.getMinutes())}`
    + `:${fillDigit(date.getSeconds())}`;
}

function createTransports() {
  return [
    new Console(Object.assign({}, DEFAULT_COMMON_OPTIONS)),
  ];
}

function colorLevel(str) {
  if(str === 'error') {
    return chalk.red(str);
  } else if(str === 'info') {
    return chalk.green(str);
  }

  return str;
}

function createLogger(level = 'warn') {
  return createWinstonLogger({
    format: combine(printf(info =>
      `${getFormatedDate()} - ${colorLevel(info.level)} ${info.message}`)),
    level,
    transports: createTransports(),
  });
}

function changeLevel(logger, level) {
  logger.configure({
    level,
    transports: createTransports(),
  });
}

const logger = createLogger();

module.exports = {
  logger,
  createLogger,
  changeLevel,
};


src/server.js
const { logger, changeLevel } = require('./logger');
const { createServer: createProxyServer } = require('./proxy_server');

const DEFAULT_OPTIONS = {
  host: '127.0.0.1',
  proxies: [{
    socks: '127.0.0.1:1080'
  }],
  proxyListReloadTimeout: 60,
  port: 8080,
};

function createServer(opts) {
  const options = Object.assign({}, DEFAULT_OPTIONS, opts);

  if(typeof options.level === 'string') {
    changeLevel(logger, options.level);
  }

  const { port, proxies, host } = options;

  // eslint-disable-next-line
  console.log(`http-proxy listening: ${host}:${port}`);
  proxies.forEach(proxy => console.log(
    `socks target: ${proxy.socks}, whitelist: '${proxy.whitelist || ''}', blacklist: '${proxy.blacklist || ''}'`));

  return createProxyServer(options).listen(port, host);
}

module.exports = {
  createServer,
};


src/proxy_server.js
// inspired by https://github.com/asluchevskiy/http-to-socks-proxy
const util = require('util');
const http = require('http');
const net = require('net');
const Socks = require('socks');
const { logger } = require('./logger');

// object class definition

function ProxyServer(options) {
  // TODO: start point
  http.Server.call(this, () => { });

  this.proxyList = [];

  if(options.proxies) {
    // stand alone proxy loging
    this.proxyList = options.proxies.map(proxy => {
      let socksProxy = parseProxyLine(proxy.socks);
      let whitelist = proxy.whitelist ? proxy.whitelist.map(regex => new RegExp(regex)) : undefined;
      let blacklist = !whitelist && proxy.blacklist ? proxy.blacklist.map(regex => new RegExp(regex)) : undefined;

      return { socksProxy, whitelist, blacklist };
    });
  }

  this.addListener(
    'request',
    requestListener.bind(null, this.proxyList)
  );
  this.addListener(
    'connect',
    connectListener.bind(null, this.proxyList)
  );
}

util.inherits(ProxyServer, http.Server);

// listeners

/**
 * called on HTTP targets
 */
function requestListener(proxyList, request, response) {
  logger.info(`request: ${request.url}`);

  const target = new URL(request.url);
  const proxy = getProxyInfo(proxyList, target);

  if(proxy) {
    onSocksHttp(target, request, response, proxy);
  }
  else {
    onNoProxyHttp(target, request, response);
  }
}

/**
 * called on HTTPs targets
 */
function connectListener(proxyList, request, socketRequest, head) {
  logger.info(`connect: ${request.url}`);

  const target = new URL(`http://${request.url}`); // connect listeners don't have the protocol in the url
  const proxy = getProxyInfo(proxyList, target);

  if(proxy) {
    onSocksHttps(target, request, socketRequest, head, proxy);
  }
  else {
    onNoProxyHttps(target, request, socketRequest, head);
  }
}

// forwarding

function onSocksHttp(target, req, res, proxy) {
  const socksAgent = new Socks.Agent({
    proxy,
    target: { host: target.hostname, port: target.port },
  });
  logger.info(`forwarding SOCKS HTTP for ${target} to ${proxy.ipaddress}:${proxy.port}`);
  sendProxyRequest(target, req, res, socksAgent);
}

function onSocksHttps(target, request, socketRequest, head, proxy) {
  const options = {
    proxy,
    target: { host: target.hostname, port: target.port },
    command: 'connect',
  };

  logger.info(`forwarding SOCKS HTTPS for ${target} to ${proxy.ipaddress}:${proxy.port}`);

  let socket;

  socketRequest.on('error', (err) => {
    logger.error(`${err.message}`);
    if(socket) {
      socket.destroy(err);
    }
  });

  Socks.createConnection(options, (error, _socket) => {
    socket = _socket;

    if(error) {
      // error in SocksSocket creation
      logger.error(`${error.message} connection creating on ${proxy.ipaddress}:${proxy.port}`);
      socketRequest.write(`HTTP/${request.httpVersion} 500 Connection error\r\n\r\n`);
      return;
    }

    socket.on('error', (err) => {
      logger.error(`${err.message}`);
      socketRequest.destroy(err);
    });

    // tunneling to the host
    socket.pipe(socketRequest);
    socketRequest.pipe(socket);

    socket.write(head);
    socketRequest.write(`HTTP/${request.httpVersion} 200 Connection established\r\n\r\n`);
    socket.resume();
  });
}

function onNoProxyHttp(target, req, res) {
  logger.info(`forwarding HTTP for ${target} without tunneling`);
  sendProxyRequest(target, req, res);
}

function onNoProxyHttps(target, req, socketRequest, head) {
  // Connect to an origin server
  logger.info(`forwarding HTTPS for ${target} without tunneling`);
  const socket = new net.Socket();

  socketRequest.on('error', (err) => {
    logger.error(`Error on request socket: ${err.message}`);
    if(socket) {
      socket.destroy(err);
    }
  });

  socket.on('error', (err) => {
    logger.error(`Error on target socket: ${err.message}`);
    socketRequest.destroy(err);
  });

  socket.connect(target.port || 80, target.hostname, () => {

    // tunneling to the host
    socket.pipe(socketRequest);
    socketRequest.pipe(socket);

    socket.write(head);
    socketRequest.write('HTTP/1.1 200 Connection Established\r\n' +
      'Proxy-agent: Node.js-Proxy\r\n' +
      '\r\n');
  });
}

function sendProxyRequest(uri, request, response, agent) {
  const options = {
    port: uri.port,
    hostname: uri.host,
    method: request.method,
    path: uri.path,
    headers: request.headers,
    agent,
  };

  const proxyRequest = http.request(options);

  request.on('error', (err) => {
    logger.error(`${err.message}`);
    proxyRequest.destroy(err);
  });

  proxyRequest.on('error', (error) => {
    logger.error(`${error.message} on connection to  ${uri.host}:${uri.port}`);
    response.writeHead(500);
    response.end('Connection error\n');
  });

  proxyRequest.on('response', (proxyResponse) => {
    proxyResponse.pipe(response);
    response.writeHead(proxyResponse.statusCode, proxyResponse.headers);
  });

  request.pipe(proxyRequest);
}

// helpers

function getProxyInfo(proxyList, target) {
  let proxyEntry = proxyList
    .find(el =>
      (!el.whitelist || regexListMatchesString(el.whitelist, target.hostname)) &&
      (!el.blacklist || !regexListMatchesString(el.blacklist, target.hostname)));

  return proxyEntry ? proxyEntry.socksProxy : undefined;

}

function regexListMatchesString(list, string) {
  return !!list.find(regEx => regEx.test(string));
}

function getProxyObject(host, port, login, password) {
  return {
    ipaddress: host,
    port: parseInt(port, 10),
    type: 5,
    authentication: { username: login || '', password: password || '' },
  };
}

function parseProxyLine(line) {
  const proxyInfo = line.split(':');

  if(proxyInfo.length !== 4 && proxyInfo.length !== 2) {
    throw new Error(`Incorrect proxy line: ${line}`);
  }

  return getProxyObject.apply(this, proxyInfo);
}

// exports

module.exports = {
  createServer: options => new ProxyServer(options),
  requestListener,
  connectListener,
  getProxyObject,
  parseProxyLine,
  getProxyInfo,
};


src/cli.js
const { readFileSync } = require('fs');
const { resolve } = require('path');
const cli = require('commander');
const { version } = require('../package.json');
const { createServer } = require('./server');

const optionNames = [
  'socks',
  'port',
  'level',
  'config',
  'host',
];

function getConfig(configValue) {

  let content = configValue;

  if(!configValue.startsWith('{')) {
    const absFile = resolve(process.cwd(), configValue);
    content = readFileSync(absFile).toString('utf8');
  }

  let fileConfig = null;

  try {
    fileConfig = JSON.parse(content);
  } catch(err) {
    const error = new Error(`invalid json content: ${err.message}`);
    error.code = err.code;
    throw error;
  }

  return fileConfig;
}

function getOptionsArgs(args) {
  const options = {};

  optionNames.forEach((name) => {
    if(Object.hasOwnProperty.apply(args, [name])) {
      if(typeof args[name] !== 'string') {
        throw new Error(`string "${name}" expected`);
      }

      if(name === 'socks') {
        options.proxies = [{ socks: args[name] }];
      }
      else {
        options[name] = args[name];
      }
    }
  });

  return options;
}

function main() {
  cli.version(version)
    .option('-s, --socks [socks]', 'specify your socks proxy hosts, default: 127.0.0.1:1080')
    .option('-p, --port [port]', 'specify the listening port of http proxy server, default: 8080')
    .option('-l, --host [host]', 'specify the listening host of http proxy server, default: 127.0.0.1')
    .option('-c, --config [config]', 'path to a json config file or the json config as string')
    .option('--level [level]', 'log level, vals: info, error')
    .parse(process.argv);

  const options = getOptionsArgs(cli.opts());

  let fileConfig = null;

  if(options.config) {
    fileConfig = getConfig(options.config);
  }

  Object.assign(options, fileConfig);

  createServer(options);
}

module.exports = {
  getOptionsArgs,
  main,
};
